expression/builtin_encryption.go

```diff
@@ -895,10 +897,160 @@ func (b *builtinUncompressedLengthSig) evalInt(row chunk.Row) (int64, bool, erro
 	return int64(len), false, nil
 }
 
+func reverse(s string) string {
+	runes := []rune(s)
+	for i, j := 0, len(runes)-1; i < j; i, j = i+1, j-1 {
+		runes[i], runes[j] = runes[j], runes[i]
+	}
+	return string(runes)
+}
+
+func validateByUser(s *variable.SessionVars, psw string) (bool, error) {
+	v, err := variable.GetGlobalSystemVar(s, variable.ValidatePasswordCheckUserName)
+	if err != nil || strings.EqualFold(v, "OFF") {
+		return err == nil, err
+	}
+	if s.User == nil {
+		return true, nil
+	}
+	if n := s.User.Username; n != "" {
+		if psw == n || psw == reverse(n) {
+			return false, nil
+		}
+	}
+	if n := s.User.AuthUsername; n != "" {
+		if psw == n || psw == reverse(n) {
+			return false, nil
+		}
+	}
+	return true, nil
+}
+
+func validateByMixedDigitSpecial(s *variable.SessionVars, pwd string) (bool, error) {
```

Comment:
Should we consider `ValidatePasswordPolicy` here?

Reply:
As I understand it, `validate_password_policy` affects what passwords pass through `validate_password`.
`VALIDATE_PASSWORD_STRENGTH` simply scores the password.

Reply:
See https://dev.mysql.com/doc/refman/8.0/en/validate-password-options-variables.html#sysvar_validate_password.policy,
`validate_password.policy` decides what test performed by `validate_password` depending on the value of the policy. 


Policy | Tests Performed
-- | --
0 or LOW | Length
1 or MEDIUM | Length; numeric, lowercase/uppercase, and special characters
2 or STRONG | Length; numeric, lowercase/uppercase, and special characters; dictionary file

Reply:
> See https://dev.mysql.com/doc/refman/8.0/en/validate-password-options-variables.html#sysvar_validate_password.policy,
> `validate_password.policy` decides what tests performed by `validate_password` depending on the value of the policy.
> 
> Policy	Tests Performed
> 0 or LOW	Length
> 1 or MEDIUM	Length; numeric, lowercase/uppercase, and special characters
> 2 or STRONG	Length; numeric, lowercase/uppercase, and special characters; dictionary file

`validate_password` is different from `VALIDATE_PASSWORD_STRENGTH`.
```
mysql> SHOW VARIABLES LIKE 'validate_password%';
+--------------------------------------+-------+
| Variable_name                        | Value |
+--------------------------------------+-------+
| validate_password_check_user_name    | ON    |
| validate_password_dictionary_file    |       |
| validate_password_length             | 8     |
| validate_password_mixed_case_count   | 1     |
| validate_password_number_count       | 1     |
| validate_password_policy             | LOW   |
| validate_password_special_char_count | 1     |
+--------------------------------------+-------+
7 rows in set (0.00 sec)

mysql> select VALIDATE_PASSWORD_STRENGTH("12345678");
+----------------------------------------+
| VALIDATE_PASSWORD_STRENGTH("12345678") |
+----------------------------------------+
|                                     50 |
+----------------------------------------+
1 row in set (0.00 sec)

mysql> select VALIDATE_PASSWORD_STRENGTH("12345678Aa@");
+-------------------------------------------+
| VALIDATE_PASSWORD_STRENGTH("12345678Aa@") |
+-------------------------------------------+
|                                       100 |
+-------------------------------------------+
1 row in set (0.00 sec)
```